framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+#ifdef CONFIG_AUDIO_MULTI_CARD
+static pthread_t g_pth_recording;
+#endif
+static sem_t g_sem_playing;
+static sem_t g_sem_recording;
+static pthread_mutex_t g_mutex_playing_q;
+static pthread_mutex_t g_mutex_recording_q;
+
+static bool g_playing_live;
+static bool g_recording_live;
+
+static dq_queue_t g_playing_q;
+static dq_queue_t g_recording_q;
+
+/***
+ * Pre-definitions
+ */
+media_result_t audio_playing(void *args);
+media_result_t audio_recording(void *args);
+
+media_result_t media_init(struct media_cb_s *cbs)
+{
+	pthread_attr_t attr;
+	pthread_mutexattr_t mutexAttr;
+	struct sched_param sparam;
+
+	// Init dq and sem
+	dq_init(&g_playing_q);
+	sem_init(&g_sem_playing, 0, 0);
+
+	pthread_mutexattr_init(&mutexAttr);
+	pthread_mutexattr_settype(&mutexAttr, PTHREAD_MUTEX_NORMAL);
+	pthread_mutex_init(&g_mutex_playing_q, &mutexAttr);
+
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	dq_init(&g_recording_q);
+	sem_init(&g_sem_recording, 0, 0);
+	pthread_mutex_init(&g_mutex_recording_q, &mutexAttr);
+#endif
+	// Init threads
+	if (pthread_attr_init(&attr) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	sparam.sched_priority = 150;
+	if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setstacksize(&attr, 8192) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_create(&g_pth_playing, &attr, (pthread_startroutine_t)audio_playing, (void *)NULL) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	if (pthread_create(&g_pth_recording, &attr, (pthread_startroutine_t)audio_recording, (void *)NULL) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+#endif
+	return MEDIA_OK;
+}
+
+void media_shutdown(void)
+{
+	// Shutdown playback thread
+	g_playing_live = false;
+	sem_post(&g_sem_playing);
+	pthread_join(g_pth_playing, NULL);
+	pthread_mutex_destroy(&g_mutex_playing_q);
+	sem_destroy(&g_sem_playing);
+
+	// Shutdown record thread
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	g_recording_live = false;
+	sem_post(&g_sem_recording);
+	pthread_join(g_pth_recording, NULL);
+	pthread_mutex_destroy(&g_mutex_recording_q);
+	sem_destroy(&g_sem_recording);
+#endif
+}
+
+media_result_t media_play(media_t *m, bool loop)
+{
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+	return MEDIA_OK;
+}
+
+media_result_t media_record(media_t *m)
+{
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pthread_mutex_lock(&g_mutex_recording_q);
+	dq_addlast((dq_entry_t *)m, &g_recording_q);
+	pthread_mutex_unlock(&g_mutex_recording_q);
+	sem_post(&g_sem_recording);
+#else
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+#endif
+	return MEDIA_OK;
+}
+
+media_result_t media_stop(media_t *m)
```

Comment:
Of course state check is mandatory for each APIs.
But if user call these apis when state is not ready, it will fail
so we should support state checking API
media_is_playing() / media_is_ready()....

Reply:
Currently we support only is_playing.

Reply:
I do not see any case where the media player will not be ready. If there is some error in some initialisation of mediaplayer , then user should handle it. 
media_is_playing is to check whether play has ended or not.

Reply:
If user want record something during play sound from cloud, it should be done immediately.
In this case state is stable everytime?

Reply:
There are two different threads , one for record and one for playback.  So the record should be able to proceed irrespective of whether a sound is playing or not. But this is only possible when we have MULTI_CARD feature in audio hardware. 
In single card, the device needs to reopened in PCM_IN mode and will have to wait until playback is complete.

Reply:
You don't understand what I mean exactly, it doesn't mean MULTI_CARD case.
I'm worried about multi-thread, because I faced many problem regarding mediaplayer's status before.
Also there are many case when application need to check mediaplayer's state.
Most simple case is turn on LED during playback or record. 
(like a play button's icon change in music player application of phone in this case application checks mediaplayer.isPlaying() )